buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,210 @@
+            val functionBuilder = StringBuilder(functionDeclaration)
+            val declarationIndex = fileText.indexOf(functionDeclaration)
+            fileText = fileText.substring(declarationIndex + functionDeclaration.length)
+            val tokens: Queue<String> = LinkedList(splitRegex.split(fileText))
+            processCurlyBlock(functionBuilder, tokens)
+            processCurlyBlock(functionBuilder, tokens)
+            functionMap.put(functionName, functionBuilder.toString())
+        }
+    }
+
+    /**
+     * Parses a block of text surrounded by curly braces. The function keeps track
+     * of the number of curly braces that have yet to be closed. Once this counter
+     * reaches 0 the block has been fully parsed. The function also ensures that
+     * curly braces within single-line comments, multi-line comments, and strings
+     * are not counted for this.
+     *
+     * @param functionBuilder The builder unto which the tokens get added to once parsed
+     * @param tokens The tokens to be parsed
+     */
+    private fun processCurlyBlock(functionBuilder: StringBuilder, tokens: Queue<String>) {
+        var curlyCounter = 0
+        var insideString = false
+        var insideSingleLineComment = false
+        var insideMultiLineComment = false
+        var validBrace = true
+        while (tokens.isNotEmpty()) {
+            val token = tokens.poll()
+            functionBuilder.append(token)
+            validBrace = !insideString &&
+                    !insideSingleLineComment &&
+                    !insideMultiLineComment
+            if (token == "\n") {
+                if (insideSingleLineComment) {
+                    insideSingleLineComment = false
+                }
+                // Allows for an abitrary amount of new lines to be parsed.
+                continue
+            }
+            if (token == "\"" && !insideSingleLineComment && !insideMultiLineComment) {
+                insideString = !insideString
+            } else if (token == "//" && !insideMultiLineComment) {
+                insideSingleLineComment = true
+            } else if (token == "/*" && !insideString && !insideSingleLineComment) {
+                insideMultiLineComment = true
+            } else if (token == "*/" && !insideString && !insideSingleLineComment) {
+                insideMultiLineComment = false
+            } else if (token == "{" && validBrace) {
+                curlyCounter++
+            } else if (token == "}" && validBrace) {
+                curlyCounter--
+            }
+            if (curlyCounter == 0) {
+                return
+            }
+        }
+    }
+
+    /**
+     * Gets the function name from the declaration.
+     *
+     * @param functionDeclaration The function declaration of form
+     *                            <return_type> <name> (<args>) :
+     */
+    private fun getFunctionName(functionDeclaration: String): String {
+        val nameRegex = Regex("\\w+")
```

Comment:
Same for this regex. Can we instantiate it in one location instead of for every function?